document-store/src/integrationTest/java/org/hypertrace/core/documentstore/mongo/MongoQueryExecutorTest.java

```diff
@@ -0,0 +1,306 @@
+    collection = datastore.getCollection(COLLECTION_NAME);
+
+    Map<Key, Document> documents = createDocumentsFromResource("mongo/collection_data.json");
+    collection.bulkUpsert(documents);
+  }
+
+  @AfterAll
+  public static void shutdown() {
+    mongo.stop();
+  }
+
+  @Test
+  public void testFindAll() throws IOException {
+    Query query = Query.builder().build();
+
+    Iterator<Document> resultDocs = collection.find(query);
+    assertSizeEqual(resultDocs, "mongo/collection_data.json");
+  }
+
+  @Test
+  public void testFindSimple() throws IOException {
+    List<SelectionSpec> selectionSpecs =
+        List.of(
+            SelectionSpec.of(IdentifierExpression.of("item")),
+            SelectionSpec.of(IdentifierExpression.of("price")),
+            SelectionSpec.of(IdentifierExpression.of("quantity")),
+            SelectionSpec.of(IdentifierExpression.of("date")));
+    Selection selection = Selection.builder().selectionSpecs(selectionSpecs).build();
+    Filter filter =
+        Filter.builder()
+            .expression(
+                RelationalExpression.of(
+                    IdentifierExpression.of("item"),
+                    NOT_IN,
+                    ConstantExpression.ofStrings(List.of("Soap", "Bottle"))))
+            .build();
+
+    Query query = Query.builder().setSelection(selection).setFilter(filter).build();
+
+    Iterator<Document> resultDocs = collection.find(query);
+    assertDocsEqual(resultDocs, "mongo/simple_filter_response.json");
+  }
+
+  @Test
+  public void testFindWithSortingAndPagination() throws IOException {
+    List<SelectionSpec> selectionSpecs =
+        List.of(
+            SelectionSpec.of(IdentifierExpression.of("item")),
+            SelectionSpec.of(IdentifierExpression.of("price")),
+            SelectionSpec.of(IdentifierExpression.of("quantity")),
+            SelectionSpec.of(IdentifierExpression.of("date")));
+    Selection selection = Selection.builder().selectionSpecs(selectionSpecs).build();
+
+    Filter filter =
+        Filter.builder()
+            .expression(
+                RelationalExpression.of(
+                    IdentifierExpression.of("item"),
+                    IN,
+                    ConstantExpression.ofStrings(List.of("Mirror", "Comb", "Shampoo", "Bottle"))))
+            .build();
+
+    Sort sort =
+        Sort.builder()
+            .sortingSpec(SortingSpec.of(IdentifierExpression.of("quantity"), DESC))
+            .sortingSpec(SortingSpec.of(IdentifierExpression.of("item"), ASC))
+            .build();
+
+    Pagination pagination = Pagination.builder().offset(1).limit(3).build();
+
+    Query query =
+        Query.builder()
+            .setSelection(selection)
+            .setFilter(filter)
+            .setSort(sort)
+            .setPagination(pagination)
+            .build();
+
+    Iterator<Document> resultDocs = collection.find(query);
+    assertDocsEqual(resultDocs, "mongo/filter_with_sorting_and_pagination_response.json");
```

Comment:
The doc id `4` with item `Shampoo` has quantity `20`, why is that on the list?
```
{
    "_id": 4,
    "item": "Shampoo",
    "price": 5,
    "quantity": 20,
    "date": "2014-04-04T11:21:39.736Z"
  }
```
`Mirror` is offset out, but `Shampoo` should have been there, right?

Reply:
We sort all the documents based on the descending order of quantity (and then, by ascending order of item).

So, after sorting, we have the following list:
`[
  {
    "_id": 4,
    "item": "Shampoo",
    "price": 5,
    "quantity": 20,
    "date": "2014-04-04T11:21:39.736Z"
  },
  {
    "_id": 7,
    "item": "Comb",
    "price": 7.5,
    "quantity": 10,
    "date": "2015-09-10T08:43:00Z",
    "props": {
      "seller": {
        "name": "Go Go Plastics",
        "address": {
          "city": "Kolkata",
          "pincode": 700007
        }
      }
    }
  },
  {
    "_id": 3,
    "item": "Shampoo",
    "price": 5,
    "quantity": 10,
    "date": "2014-03-15T09:00:00Z",
    "props": {
      "brand": "Sunsilk",
      "size": "L",
      "seller": {
        "name": "Metro Chemicals Pvt. Ltd.",
        "address": {
          "city": "Mumbai",
          "pincode": 400004
        }
      }
    }
  },
  {
    "_id": 6,
    "item": "Comb",
    "price": 7.5,
    "quantity": 5,
    "date": "2015-06-04T05:08:13Z"
  },
  {
    "_id": 8,
    "item": "Soap",
    "price": 10,
    "quantity": 5,
    "date": "2016-02-06T20:20:13Z"
  },
  {
    "_id": 5,
    "item": "Soap",
    "price": 20,
    "quantity": 5,
    "date": "2014-04-04T21:23:13.331Z",
    "props": {
      "brand": "Lifebuoy",
      "size": "S",
      "seller": {
        "name": "Hans and Co.",
        "address": {
          "city": "Kolkata",
          "pincode": 700007
        }
      }
    }
  },
  {
    "_id": 1,
    "item": "Soap",
    "price": 10,
    "quantity": 2,
    "date": "2014-03-01T08:00:00Z",
    "props": {
      "brand": "Dettol",
      "size": "M",
      "seller": {
        "name": "Metro Chemicals Pvt. Ltd.",
        "address": {
          "city": "Mumbai",
          "pincode": 400004
        }
      }
    }
  },
  {
    "_id": 2,
    "item": "Mirror",
    "price": 20,
    "quantity": 1,
    "date": "2014-03-01T09:00:00Z"
  }
]
`
From this sorted list, we are applying pagination with offset 1 and limit 3. Hence, the result in "filter_with_sorting_and_pagination_response.json".
`Mirror` is last in the list so there is no way to get that since we have a limit 3.

The document in question
`{
    "_id": 4,
    "item": "Shampoo",
    "price": 5,
    "quantity": 20,
    "date": "2014-04-04T11:21:39.736Z"
  }` is not present since it is first in the sorted list and we use an offset of 1.